biosimulators_utils/model_lang/smoldyn/simularium_converter.py

```diff
@@ -0,0 +1,293 @@
+"""Using the Biosimulators side of Smoldyn to generate a modelout.txt Smoldyn file for a specified OMEX/COMBINE archive which then
+    is used to generate a .simularium file for the given simulation. That .simularium file is then stored along with the log.yml and
+    report.{FORMAT} relative to the simulation. Remember: each simulation, while not inherently published, has the potential for publication
+    based purely on the simulation's ability to provide a valid OMEX/COMBINE archive. There exists (or should exist) an additional layer
+    of abstraction to then validate and verify the contents therein.
+"""
+
+
+import os
+from typing import Optional, Tuple, Dict, List, Union
+from abc import ABC, abstractmethod
+import numpy as np
+import pandas as pd
+from simulariumio import (
+    CameraData,
+    UnitData,
+    MetaData,
+    DisplayData,
+    DISPLAY_TYPE,
+    BinaryWriter,
+)
+from simulariumio.smoldyn.smoldyn_data import InputFileData
+from simulariumio.smoldyn import SmoldynConverter, SmoldynData
+from simulariumio.filters import TranslateFilter
+from simulariumio.data_objects.trajectory_data import TrajectoryData, AgentData
+
+
+__all__ = [
+    'CombineArchive',
+    'BiosimulatorsDataConverter',
+    'SmoldynDataConverter',
+]
+
+
+class CombineArchive:
+    def __init__(self,
+                 rootpath: str,
+                 outputs_dirpath: Optional[str] = None,
+                 model_output_filename: Optional[str] = None,
+                 simularium_filename: Optional[str] = None,
+                 name='my_combine_archive'):
+        self.rootpath = rootpath
+        self.outputs_dirpath = outputs_dirpath
+        self.simularium_filename = simularium_filename or f'{name}_output_for_simularium'
+        self.paths = self.__get_all_archive_filepaths()
+        self.model_path = self.set_model_filepath()
+
+        self.model_output_filename = model_output_filename \
+            or self.model_path.replace('.txt', '') + 'out.txt'
+        self.paths['model_output_file'] = self.model_output_filename
+
+    def __get_all_archive_filepaths(self) -> Dict[str, str]:
+        paths = {}
+        if os.path.exists(self.rootpath):
+            for root, _, files in os.walk(self.rootpath):
+                paths['root'] = root
+                for f in files:
+                    fp = os.path.join(root, f)
+                    paths[f] = fp
+        return paths
+
+    def set_model_filepath(self, model_filename: Optional[str] = None) -> Union[str, None]:
+        model_filename = model_filename or 'model.txt'  # default Smoldyn model name
+        for k in self.paths.keys():
+            full_path = self.paths[k]
+            if model_filename in full_path:
+                return full_path
+
+
+class BiosimulatorsDataConverter(ABC):
+    def __init__(self, archive: CombineArchive):
```

Comment:
Is this abstract class designed to be general or only used in Smoldyn?

Reply:
This is an abstract class with SmoldynDataConverter as one of the implementations. This is meant to be an abstract Data converter that will be implemented as more simularium-friendly biosimulators become available.